src/fiber/fiber.ml

```diff
@@ -1,243 +1,455 @@
     type 'a t
 
-    (* Create a continuation that captures the current execution context *)
-    val create : ('a -> unit) -> 'a t
+    val create : unit -> 'a t
+
+    val read : 'a t -> 'a fiber
+
+    val fill : 'a t -> 'a -> unit fiber
 
-    (* Enqueue the restarting of a suspended fiber. *)
-    val enqueue : 'a t -> 'a -> unit
+    val peek : 'a t -> 'a option fiber
   end
 
-  (* Execute a function returning a fiber, passing any raised exception to the
-     current execution context. This function preserve the current execution
-     context. It should be called when creating forks.*)
-  val apply : ('a -> 'b t) -> 'a -> 'b t
+  module Var : sig
+    type 'a t
+
+    val create : unit -> 'a t
 
-  val apply2 : ('a -> 'b -> 'c t) -> 'a -> 'b -> 'c t
+    val get : 'a t -> 'a option fiber
 
-  (* Add [n] references to the current execution context *)
-  val add_refs : int -> unit
+    val get_exn : 'a t -> 'a fiber
 
-  (* Decrease the reference count of the current execution context *)
-  val deref : unit -> unit
+    val set : 'a t -> 'a -> (unit -> 'b fiber) -> 'b fiber
 
-  (* [wait_errors f] executes [f ()] inside a new execution contexts. Returns a
-     fiber that terminates when all the fiber in the sub-context have
-     terminated. *)
-  val wait_errors : (unit -> 'a t) -> ('a, unit) result t
+    val unset : 'a t -> (unit -> 'b fiber) -> 'b fiber
+  end
 
-  (* Set the current error handler. [on_error] is called in the current
-     execution context. *)
-  val set_error_handler :
-    on_error:(Exn_with_backtrace.t -> unit t) -> ('a -> 'b t) -> 'a -> 'b t
+  type 'a k
 
-  val vars : unit -> Univ_map.t
+  (* Suspend the execution of a fiber *)
+  val suspend : ('a k -> unit) -> 'a fiber
 
-  val set_vars : Univ_map.t -> ('a -> 'b t) -> 'a -> 'b t
+  (* Restart a suspended fiber. Must only be called once on a given suspended
+     fiber *)
+  val restart : 'a k -> 'a -> unit fiber
 
-  val run : 'a t -> iter:(unit -> unit) -> 'a
+  type fill = Fill : 'a Ivar.t * 'a -> fill
 
-  val reraise_all : Exn_with_backtrace.t list -> unit
+  val run : 'a t -> iter:(unit -> fill Nonempty_list.t) -> 'a
 end = struct
-  type t =
-    { on_error : Exn_with_backtrace.t k option
-          (* This handler must never raise *)
+  type 'a t = ('a -> effect) -> effect
+
+  and effect =
+    | Read_ivar : 'a ivar * ('a -> effect) -> effect
+    | Fill_ivar : 'a ivar * 'a * (unit -> effect) -> effect
+    | Suspend : ('a k -> unit) * ('a -> effect) -> effect
+    | Restart : 'a k * 'a * (unit -> effect) -> effect
+    | Get_var : 'a Univ_map.Key.t * ('a option -> effect) -> effect
+    | Set_var :
+        'a Univ_map.Key.t * 'a * (unit -> 'b t) * ('b -> effect)
+        -> effect
+    | Unset_var : 'a Univ_map.Key.t * (unit -> 'b t) * ('b -> effect) -> effect
+    | With_error_handler :
+        (unit -> 'a t) * (Exn_with_backtrace.t -> Nothing.t t) * ('a -> effect)
+        -> effect
+    | Unwind : context * ('a -> effect) * 'a -> effect
```

Comment:
Could you leave comments for some of the non-obvious constructors. Namely Suspend, Restart, Unwind.